### ISSUE TITLE
Dask cluster worker specs based on task resource requirements

### ISSUE THREAD

**Issue by AlexanderVanEck** [PrefectHQ/prefect#4705]
### Discussed in https://github.com/PrefectHQ/prefect/discussions/4700

<div type='discussions-op-text'>

<sup>Originally posted by **AlexanderVanEck** June 25, 2021</sup>
Hello! 👋  Thank you for creating/maintaining prefect, I've been exploring it for the past week and so far it's been wonderful!
and I think I've stumbled upon an interesting feature/idea that we might consider. Please let me know if this is already currently possible somehow, I'm not an prefect expert (yet) ;)

### Problem hypothesis

Let's say I have the following task/flows;

```
import numpy as np
import prefect
import torch
from prefect import Flow, task
from prefect.executors import DaskExecutor


@task(tags=['dask-resources:GPU=1'])
def a_gpu_task():
    logger = prefect.context.get("logger")
    logger.info(f"Hello, Torch CUDA {torch.cuda.is_available()}!")


@task(tags=['dask-resources:MEMORY=20e9'])  # 20GB
def a_high_memory_task():
    np.ones(10_000_000_000, dtype=np.uint64)  # 10GB


@task
def a_cpu_task():
    logger = prefect.context.get("logger")
    logger.info(f"Hello, Cloud!")


with Flow(
    'sample-flow',
    executor=DaskExecutor(
        cluster_class="dask_cloudprovider.aws.FargateCluster",
        cluster_kwargs={"n_workers": 3, "image": "my-prefect-image"},
    ),
) as flow:
    a_gpu_task()
    a_high_memory_task()
    a_cpu_task()

```

The DaskExecutor here will reach out to AWS and create a FargateCluster. My hypothesis is that since no worker resource requirements are provided in `cluster_kwargs` the gpu/memory tasks will never schedule/complete. (assume that the config.yaml also doesn't set requirements). Is that correct?

### The documented way

We can obviously include worker specs to the cluster_kwargs and make sure workers have the right amount of resources.

```
with Flow(
    'sample-flow',
    executor=DaskExecutor(
        cluster_class="dask_cloudprovider.aws.FargateCluster",
        cluster_kwargs={
            "n_workers": 3,
            "image": "my-prefect-image",
            "worker_gpu": 1,
            "worker_mem": 20e9,
        },
    ),
) as flow:
    a_gpu_task()
    a_high_memory_task()
    a_cpu_task()
```

This solves the blocking issue and creates 3 workers with a GPU each and high-memory enough memory.

A couple of thoughts on this behavior:

1. We push it onto the user to find an acceptable worker spec that satisfies all task resource needs. Every time a task is added/removed from a a flow the user will have to update/confirm the worker spec.
2. This creates 3 workers with the highest resources possible, GPUs attached to workers that aren't going to use them and high-memory everywhere. I think we can agree that this is not the most optimal dask worker configuration for this task workload.
3. All 3 workers use the same docker image even though their runtimes may be different, both in dependencies installed and dask-worker process used (f.e. the gpu task could use `dask-cuda-worker` instead of `dask-worker`).

### How to make this work?

If this cannot be expressed today I think it could be possible to make it work as follows;

1. Create a JITDaskExecutor.

* Spins up no workers during startup, only scheduler
* Before submitting a task scales up a worker
* Task tags can be used to specify requirements of runner other cluster defaults are taken
* After task is complete worker is scaled down again 

### Questions

1. Is this is even how Dask is supposed to work?

I see that `SpecCluster` can support several worker specs through the `workers` attribute. but I haven't tested it. and I see almost no support for this on any other Cluster objects.

2. How many dask Cluster objects actually have methods to cleanly spin up different workers? dask-jobqueue f.e. doesn't seem to have this ability.

3. Are there any alternatives to this?

How are we doing resource optimization at the moment? Given that the above stated problem can be regarded as wasteful?

I'd love to hear your thoughts on if this is a viable approach or alternative approaches!

It seems Dask has also been speaking about this: https://github.com/dask/distributed/issues/2118

</div>

@joshmeek Would love to hear your thoughts on this. :)

**Comment by madkinsz** [comment]
Hey @AlexanderVanEck 

Thanks for putting together this user-story! Unfortunately, this isn't really something we can support as it's blocked by support in dask. Managing resources is a bit out of scope for Prefect right now and we'll have to rely on implementations upstream.

To answer your question
> How are we doing resource optimization at the moment?

I think most users that care about this are using autoscaling kubernetes clusters and ecs tasks. Generally, the flow run is the unit of computation that we support resource management for. At the task level, it gets tricky to support across all of our execution environments.

I'll close this issue for now as most discussion related to this will have to happen in dask.

**Comment by AlexanderVanEck** [comment]
Thank you for the explanation @madkinsz. 🙌 

Could you also close this question? https://github.com/PrefectHQ/prefect/discussions/4700

**Comment by madkinsz** [comment]
No problem! Definitely feel free to ask more questions :) resource management is a bit of a tricky area--it is certainly an important thing, but it is at the edge of what we can reasonably address as a tool.

**Comment by mrocklin** [comment]
Jumping in here.  Dask certainly supports multiple worker types within one cluster, but as noted by @AlexanderVanEck almost none of the implementations make this easy.  

My experience has been that it's a rare enough collection of cases, each of which has slightly different expectations that it's been hard to build a consistent API around it that pleases everyone (or anyone, really).  The lower level system `SpecCluster` is designed to accomodate this flexibility, but it's really more developer focused rather than user focused.  

If Prefect presents a unified set of requirements around this then let's dig in and find some way to make spelling this out easier for users.    My *guess* though is that we'll have to specialize it to a specific deployment strategy and a specific use case though.  Specifying architecture is hard :/
